ext/webserial/lib.rs

```diff
@@ -0,0 +1,217 @@
+      ("op_webserial_open_port", op_sync(op_webserial_open_port)),
+      ("op_webserial_read", op_async(op_webserial_read)),
+      ("op_webserial_write", op_async(op_webserial_write)),
+      (
+        "op_webserial_set_signals",
+        op_async(op_webserial_set_signals),
+      ),
+      (
+        "op_webserial_get_signals",
+        op_async(op_webserial_get_signals),
+      ),
+    ])
+    .build()
+}
+
+pub fn get_declaration() -> PathBuf {
+  PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("lib.deno_webserial.d.ts")
+}
+
+pub struct SerialPortResource(AsyncRefCell<tokio_serial::SerialStream>);
+
+impl Resource for SerialPortResource {
+  fn name(&self) -> Cow<str> {
+    "serialPort".into()
+  }
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SerialPortInfo {
+  name: String,
+  info: Option<(u16, u16)>,
+}
+
+pub fn op_webserial_get_ports(
+  _: &mut OpState,
+  _: (),
+  _: (),
+) -> Result<Vec<SerialPortInfo>, AnyError> {
+  let ports = tokio_serial::available_ports()?
+    .into_iter()
+    .map(|info| SerialPortInfo {
+      name: info.port_name,
+      info: match info.port_type {
+        serialport::SerialPortType::UsbPort(info) => Some((info.vid, info.pid)),
+        _ => None,
+      },
+    })
+    .collect::<Vec<SerialPortInfo>>();
+  Ok(ports)
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct OpenPortArgs {
+  baud_rate: u32,
+  data_bits: u8,
+  stop_bits: u8,
+  parity: String,
+  flow_control: String,
+}
+
+pub fn op_webserial_open_port(
+  state: &mut OpState,
+  name: String,
+  args: OpenPortArgs,
+) -> Result<ResourceId, AnyError> {
+  let serial_port = tokio_serial::new(&name, args.baud_rate)
+    .data_bits(match args.data_bits {
+      7 => tokio_serial::DataBits::Seven,
+      8 => tokio_serial::DataBits::Eight,
+      _ => unreachable!(),
+    })
+    .stop_bits(match args.stop_bits {
+      1 => tokio_serial::StopBits::One,
+      2 => tokio_serial::StopBits::Two,
+      _ => unreachable!(),
+    })
+    .parity(match args.parity.as_str() {
+      "none" => tokio_serial::Parity::None,
+      "even" => tokio_serial::Parity::Even,
+      "odd" => tokio_serial::Parity::Odd,
+      _ => unreachable!(),
+    })
+    .flow_control(match args.flow_control.as_str() {
+      "none" => tokio_serial::FlowControl::None,
+      "hardware" => tokio_serial::FlowControl::Hardware,
+      _ => unreachable!(),
+    })
+    .open_native_async()?;
+
+  let rid = state
+    .resource_table
+    .add(SerialPortResource(AsyncRefCell::new(serial_port)));
+
+  Ok(rid)
+}
+
+pub async fn op_webserial_read(
+  state: Rc<RefCell<OpState>>,
+  rid: ResourceId,
+  mut buf: ZeroCopyBuf,
+) -> Result<(), AnyError> {
+  let resource = state
+    .borrow_mut()
+    .resource_table
+    .get::<SerialPortResource>(rid)?;
+  let mut serial_port = RcRef::map(resource, |r| &r.0).borrow_mut().await;
+  serial_port.read_exact(buf.as_mut()).await?;
```

Comment:
Shouldn't we just use `read` instead of `read_exact` here? Buffer size is determined by `controller.desiredSize` but what if there's not enough data in serial port to read, would the stream just not pull until there is data of desired size?

Reply:
Again. This PR is not anywhere near being ready.